app/javascript/components/remove-catalog-item-modal.jsx

```diff
@@ -0,0 +1,142 @@
+    super(props);
+    this.state = {
+      data: []
+    };
+  }
+
+  componentDidMount() {
+    let apiPromises = [];
+    const catalogItemsIds = ManageIQ.record.recordId ? [ManageIQ.record.recordId] : ManageIQ.gridChecks;
+
+    // Load modal data from API
+    catalogItemsIds.map(item => {
+      apiPromises.push(API.get(`/api/service_templates/${item}?attributes=services`));
+    });
+    Promise.all(apiPromises)
+      .then((apiData) => (apiData.map(catalogItem => (
+        {id:           catalogItem.id,
+         name:         catalogItem.name,
+         service_type: catalogItem.service_type, // 'atomic' or 'composite'
+         services:     catalogItem.services}))))
+      .then((data) => this.setState({data: data}));
+
+    // Buttons setup
+    this.props.dispatch({
+      type: 'FormButtons.init',
+      payload: {
+        newRecord: true,
+        pristine: true,
+        addClicked: () => removeCatalogItems(this.state.data)
+      }
+    });
+    this.props.dispatch({
+      type: "FormButtons.customLabel",
+      payload: __('Remove'),
+    });
+    this.props.dispatch({
+      type: 'FormButtons.saveable',
+      payload: true
+    });
+  }
+
+  render () {
+    const usedServicesMessage = (data) => {
+      let warningItems = [];
+      if (data.length === 1) { // We're deleting just one catalog item
+        warningItems = data[0].services.map(service => ({id: service.id, name: service.name}));
+      } else {                 // We're deleting multiple catalog items
+        warningItems = data.filter(item => item.services && item.services.length > 0);
+      }
+      if (warningItems.length > 0) {
+        let warningMessage = '';
+        if (data.length === 1 && isCatalogBundle(data[0])) {
+          warningMessage = __('The catalog bundle is linked to the following services:');
+        } else {
+          warningMessage = n__('The catalog item is linked to the following services:',
+                               'The following catalog items are linked to services:', data.length)
+        }
+        return (
+          <div>
+            <h4>{warningMessage}</h4>
+            {warningItems.map(item => (
+              <ul key={item.id}><h4><strong>{item.name}</strong></h4></ul>
+            ))}
+          </div>
+        );
+      }
+    };
+
+    const confirmationMessage = (data) => {
+      if (data.length === 1 && isCatalogBundle(data[0])) {
+        return __('Are you sure you want to remove the following catalog bundle?');
+      } else
+      {
+        return n__('Are you sure you want to remove the following catalog item?',
+                   'Are you sure you want to remove the following catalog items?', data.length);
+      }
+    };
+
+    return (
+      <Modal.Body className="warning-modal-body">
+        {this.state.data && usedServicesMessage(this.state.data)}
+        <div>
+           <h4>{confirmationMessage(this.state.data)}</h4>
+           {this.state.data && this.state.data.map(item => (
```

Comment:
Is it possible that there are no data in state? You are setting it up in constructor and api should return something always?

Reply:
Fixed.